src/state-manager/TransactionQueue.ts

```diff
@@ -411,15 +435,63 @@ class TransactionQueue {
 
           for (let i = 0; i < req.stateList.length; i++) {
             // eslint-disable-next-line security/detect-object-injection
-            const state = req.stateList[i]
-            if (
-              i !== 0 &&
-              !this.validateCorrespondingTellSender(queueEntry, state.accountId, header.sender_id)
-            ) {
-              /* prettier-ignore */ if (logFlags.error && logFlags.verbose) this.mainLogger.error(`${route} validateCorrespondingTellSender failed for ${state.accountId}`)
-              return errorHandler(RequestErrorEnum.InvalidSender)
+            const state = req.stateList[i];
+            let isSenderValid = false
+            const senderNodeId = header.sender_id
+            if (configContext.p2p.useFactCorrespondingTell) {
+              let isSenderOurExeNeighbour = false
+              // check if it is a neighbour exe node sharing data
+              if (configContext.stateManager.shareCompleteData) {
+                const senderIsInExecutionGroup = queueEntry.executionGroupMap.has(senderNodeId)
+                const neighbourNodes = utils.selectNeighbors(queueEntry.executionGroup, queueEntry.ourExGroupIndex, 2) as Shardus.Node[]
+                const neighbourNodeIds = neighbourNodes.map((node) => node.id)
+                isSenderOurExeNeighbour = senderIsInExecutionGroup && neighbourNodeIds.includes(senderNodeId)
+                if (isSenderOurExeNeighbour) {
+                  nestedCountersInstance.countEvent('stateManager', 'factValidateCorrespondingTellSender: sender is an execution node and a neighbour node')
+                  isSenderValid = true
+                } else {
+                  // check if it is a corresponding tell sender
+                  isSenderValid = this.factValidateCorrespondingTellSender(
+                    queueEntry,
+                    state.accountId,
+                    senderNodeId
+                  )
+                }
+              } else {
+                // check if it is a corresponding tell sender
+                isSenderValid = this.factValidateCorrespondingTellSender(
+                  queueEntry,
+                  state.accountId,
+                  senderNodeId
+                )
+              }
+            } else {
+              isSenderValid = this.validateCorrespondingTellSender(queueEntry, state.accountId, senderNodeId)
+            }
+
+            if (this.stateManager.testFailChance(configContext.debug.ignoreDataTellChance, 
+            'ignoreDataTellChance', queueEntry.logID, '', logFlags.verbose ) === true ) {
+              isSenderValid = false
+            }
+
+            if (i !== 0 && isSenderValid === false) {
```

Comment:
what is the logic under i !== 0 condition ? why senderinvalid is only important when it is second iteration?

Reply:
oh wow.   I have removed this in a local commit.   will also send to blue team for comments.  
I agree that we should also let the check apply for i === 0

Reply:
I don't remember exactly but i === 0 seems to be from very old interface where the first item is app data hash (or maybe global data?). I am not sure. But it should be 100% changed to `if (isSenderValid === false)` now.